sql/schemachanger: create function declarative

chengxiong-ruan opened the pull request:
**sql/schemachanger: rename UpdateBackReferencesInSequences**

  Rename UpdateBackReferencesInSequences to
  UpdateTableBackReferencesInSequences since it's fits more
  specifically for tables referencing sequences.

**sql/schemachanger: resolve schema should check required privilege**

  Previously, we require current user's ownership on a schema
  descriptor when resolving a schema in declarative shcema
  changer. This was true because we only support `DROP SCHEMA`.
  But this won't be true if we want to support `CREATE`s because
  we'd only require `CREATE` privilege on the schema. This pr
  adds a new `RequireOwnership` flag to `ResolveParam` since there
  is no `OWNER` privilege kind.

**sql/schemachanger: Add default privileges elements and grant option field**

  This commit adds a new element for default privileges. Also adds
  a grant option field to UserPrivielges element.

**sql/schemachanger: add id generator and references provider interface**

  This commit adds `ReferenceProviderFactory` and `DescIDGenerator`
  interfaces to help doing CREATEs within declarative schema changer.

**sql/schemachanger: make builder able to work with new descriptors**

  Previously, builder always try to read descriptors because we
  assume all descriptors already exist. This is not true anymore
  for `CREATE` statements. This commit adds a new id set to the
  builder so it knows when to read a descriptor to initilize the
  cache or not.

**sql/schemachanger: create function declarative**

  This commits implements CREATE FUNCTION statements in the
  declarative schema changer.

  It introduces a few new things:
  (1) A new `DESCRIPTOR_ADDING` state for a newly created descritpor.
  In this state the descriptor is visible only in a cache of
  `mutationVisitorState` so that other operations in the same statement
  can see the new descriptor. At this state, descriptor is not written
  to collection and storage yet.
  (2) Introduce operations to update `ObjectParent`, `Owner` and
  `UserPrivileges` elements.
  (3) add helper to able to create user privileges from default privileges
  of database and schemas.

  The rest of the commit is mainly implementing the business logic.

Fixes: #83228

chengxiong-ruan committed:
sql: resolve schema should check required privilege

Previously, we require current user's ownership on a schema
descriptor when resolving a schema in declarative shcema
changer. This was true because we only support `DROP SCHEMA`.
But this won't be true if we want to support `CREATE`s because
we'd only require `CREATE` privilege on the schema. This pr
adds a new `RequireOwnership` flag to `ResolveParam` since there
is no `OWNER` privilege kind.

Epic: None

Release note: None.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/93906)
<!-- Reviewable:end -->

chengxiong-ruan committed:
schemachanger: rename UpdateBackReferencesInSequences

Rename UpdateBackReferencesInSequences to
UpdateTableBackReferencesInSequences since it's fits more
specifically for tables referencing sequences.

Epic: None
Release note: None

chengxiong-ruan committed:
sql: add grant option field

This commit adds a grant option field to UserPrivielges element.

Release note: None.

chengxiong-ruan committed:
sql/schemachanger: add id generator and references provider interface

This commit adds `ReferenceProviderFactory` and `DescIDGenerator`
interfaces to help doing CREATEs within declarative schema changer.

Release note: None

chengxiong-ruan committed:
sql/schemachanger: make builder able to work with new descriptors

Previously, builder always try to read descriptors because we
assume all descriptors already exist. This is not true anymore
for `CREATE` statements. This commit adds a new id set to the
builder so it knows when to read a descriptor to initilize the
cache or not.

Release note: None

chengxiong-ruan committed:
sql/schemachanger: create function declarative

This commits implements CREATE FUNCTION statements in the
declarative schema changer.

It introduces a few new things:
(1) A new `DESCRIPTOR_ADDING` state for a newly created descritpor.
In this state the descriptor is visible only in a cache of
`mutationVisitorState` so that other operations in the same statement
can see the new descriptor. At this state, descriptor is not written
to collection and storage yet.
(2) Introduce operations to update `ObjectParent`, `Owner` and
`UserPrivileges` elements.
(3) add helper to able to create user privileges from default privileges
of database and schemas.

The rest of the commit is mainly implementing the business logic.

Release note: None

chengxiong-ruan commented:
There're a lot of conflict after [#95786](https://github.com/cockroachdb/cockroach/pull/95786) was merged, and I assume there will be more conflict with the important work in #95631 . I would prefer rebase after it and I think the main things are ready for a review.

postamar commented on pkg/sql/schemachanger/scpb/elements.proto:
```diff
@@ -632,7 +635,6 @@ message FunctionBody {
   repeated ViewReference uses_views = 5 [(gogoproto.nullable) = false];
   repeated uint32 uses_sequence_ids = 6 [(gogoproto.customname) = "UsesSequenceIDs", (gogoproto.casttype) = "github.com/cockroachdb/cockroach/pkg/sql/sem/catid.DescID"];
   repeated uint32 uses_type_ids = 7 [(gogoproto.customname) = "UsesTypeIDs", (gogoproto.casttype) = "github.com/cockroachdb/cockroach/pkg/sql/sem/catid.DescID"];
-  repeated uint32 uses_function_ids = 8 [(gogoproto.customname) = "UsesFunctionIDs", (gogoproto.casttype) = "github.com/cockroachdb/cockroach/pkg/sql/sem/catid.DescID"];
```
Is this deliberate?

postamar commented on pkg/sql/schemachanger/scpb/elements.proto:
```diff
@@ -643,3 +645,12 @@ message FunctionParamDefaultExpression {
+    cockroach.sql.sqlbase.DefaultPrivilegesForRole.ForAllRolesPseudoRole for_all_role = 3;
+  }
+  UserPrivileges user_privileges = 4 [(gogoproto.nullable) = false, (gogoproto.embed) = true];
+}
```
I don't know about this. I feel that it's too early to introduce this element because it's unclear how to model it correctly. Also, we don't actually define any schema changes on it, it's just used for the builder. So, how about this for the time being: add a method somewhere in the scbuildstmt dependencies that returns the default privileges given an ID and whatever else it needs. You can plumb that to the `descCache` in the `builderState` somehow.

postamar replied:
For a start it's not clear how to key this element, nor whether this element should be exploded into several or not (one for explicit roles, another for all roles, etc on many dimensions). Should all privileges be bundled together?

Consider:
```
BEGIN;
ALTER DEFAULT PRIVILEGES FOR ALL ROLES IN SCHEMA public GRANT INSERT ON TABLES TO CURRENT_USER;
ALTER DEFAULT PRIVILEGES FOR ALL ROLES IN SCHEMA public GRANT DELETE ON TABLES TO CURRENT_USER;
COMMIT;
```
What would the targets look like?

It's not obvious. Until we figure this kind of thing out and exercise it in prod, let's stick to exposing this as dependencies.

As a meta-commentary I'd like to stress that this further reinforces my strong belief that modelling elements should only be done once there's a comprehensive set of schema changes operating on them, and not sooner.

chengxiong-ruan replied:
I added a `BuildUserPrivilegesFromDefaultPrivileges` which generates the new owner and user privileges elements from db and schema's default privileges. I also deleted all the code I added to reason privileges from default privilege elements (which are not gone) since we can just reuse the interesting function in `catprivilege`.

postamar commented on pkg/sql/schemachanger/scbuild/internal/scbuildstmt/dependencies.go:
```diff
@@ -329,3 +331,19 @@ type NameResolver interface {
+	// ForEachTypeReference iterate through all referenced type ids with the given
+	// function.
+	ForEachTypeReference(func(typeID descpb.ID) error) error
+}
```
We never, ever want to reason in terms of descriptors in `scbuildstmt`. Only elements. Perhaps your interface should be in terms of those validate* functions that you defined that consume this ReferenceProvider?

chengxiong-ruan replied:
This is absolutely rings a bell! Thanks for pointing it out. These methods are also need for building cross references besides validation for defending cross-db references. I would just remove the reference to table, instead just count on `table id` and `table type (table/view/sequence)` since that's all we need at the moment.

chengxiong-ruan replied:
Done.

postamar commented on pkg/sql/schemachanger/scbuild/internal/scbuildstmt/dependencies.go:
```diff
@@ -329,3 +331,19 @@ type NameResolver interface {
 }
+
+type DescriptorIDGenerator interface {
+	GenerateUniqueDescID() catid.DescID
```
This should be a method on the `BuilderState` interface. Its implementation could do whatever `MarkNewDescriptor` does so that that method never needs to be exposed in `scbuildstmt`. We really want to keep `scbuildstmt` as tight as possible and only have business logic that turns DDL statement ASTs into element targets, otherwise it very quickly becomes a huge mess.

chengxiong-ruan replied:
Totally agreed. Will do.

chengxiong-ruan replied:
change as suggested.

postamar commented on pkg/sql/schemachanger/scpb/scpb.proto:
```diff
@@ -67,6 +67,10 @@ enum Status {
+
+  // When a basic descriptor (the object element) is first created before all
+  // the other dependent elements become public.
+  DESCRIPTOR_ADDING = 23;
```
ADDED?

chengxiong-ruan replied:
Yeah, sounds better.

chengxiong-ruan replied:
done

postamar commented on pkg/sql/schemachanger/scplan/internal/opgen/opgen_function.go:
```diff
@@ -13,21 +13,32 @@ package opgen
+					}
+				}),
+			),
 			to(scpb.Status_PUBLIC,
```
This needs to be `revertible(false)`. Or perhaps the previous transition. See what makes sense by looking into something like
```
BEGIN;
CREATE FUNCTION f ...
CREATE UNIQUE INDEX idx ON t(x);
COMMIT;
```
You want the function to become visible to the rest of the cluster only once the schema change reaches the PostCommitNonRevertable phase.

chengxiong-ruan replied:
Good call, will adjust and add test case for that.

chengxiong-ruan replied:
done. though looks like there isn't a way to add such test since we still fall back within explicit transaction?

ajwerner replied:
If you set the session variable you can do it. `SET use_declarative_schema_changer = unsafe_always`.

chengxiong-ruan replied:
that is very true.

chengxiong-ruan replied:
Added a test for this. Awesome work on the execution-time generalization work. This is exciting.

chengxiong-ruan replied:
hmm.. Actually I should also have a test with the job paused instead of fail and rollback. But I found that reveal some problems here.
Like..we should also make setting ObjectParent (which add the function to schema) not revertible, so that the function won't be visible to function resolution. But we cannot do that, because it would failed pre-commit phase due to cross-reference validation on function descriptor (like it cannot have invalid parent schema). We definitely need the descriptor to exists so that we can track descriptors in jobs with no problem. I think one comprise I can do is to have the descriptor's state set to `DROP` when it's first created. The default value of `PUBLIC` doesn't make sense in anymore.

ajwerner replied:
Does the existing data-driven framework not implement this? Like if you add an `end_to_end` test, I think it can have multiple statements. Also, if you do that, I'm pretty certain that we'll generate a `pause` case where it'll pause and resume between each of them.

Maybe one thing we need to do is find a way to talk about operations inside the same transaction between the statements. @fqazi do you have thoughts?

chengxiong-ruan replied:
thanks, I added those auto generated test including the explains(_verbose) tests for multi-statements. It seems good. The explain on the rollback stages are awesome that it shows rollback on descriptors affected by the multiple statements.
I also set the initial state of the descriptor as `ADD`. I think from now being, in declarative schema changer, we need to do the same when creating other descriptors, so I refactor the `newAddingTableError` thing to be not just for table. This solves the cross-reference validation errors I saw earlier.

postamar commented on pkg/sql/schemachanger/scplan/internal/rules/current/dep_create_function.go:
```diff
@@ -0,0 +1,65 @@
+	registerDepRule(
+		"function existence precedes function dependents",
+		scgraph.Precedence,
+		"function", "dependent",
```
Should this rule be more general and apply to all descriptors and their dependents?

chengxiong-ruan replied:
made it as a general rule

postamar commented on pkg/sql/schemachanger/scplan/internal/rules/current/dep_create_function.go:
```diff
@@ -0,0 +1,65 @@
+	registerDepRule(
+		"function dependents exist before function becomes public",
+		scgraph.Precedence,
+		"dependent", "function",
```
Same comment as above.

postamar commented on pkg/sql/schemachanger/scplan/internal/rules/current/dep_create_function.go:
```diff
@@ -0,0 +1,65 @@
+	registerDepRule(
+		"function name should be set before parent ids",
+		scgraph.Precedence,
+		"function-name", "function-parent",
```
This one is probably truly function-specific, though.

postamar reviewed (changes requested):
Nice work. This can't have been easy.

I did a review pass and I have a couple of major concerns that I'd like addressed before doing another pass:
- the default privilege element needs to be replaced by dependencies in scbuildstmt, otherwise we're going to be in trouble when it's time to implement default privilege changes proper;
- the dependencies you added in scbuildstmt need to be at a higher level of abstraction, we especially shouldn't introduce descriptors in that package, ever.

More generally I think you should add more test cases.

chengxiong-ruan commented:
@postamar this is RFAL!

postamar reviewed (approved):
Very nice. I only have very minor comments which should be straightforward to address.

chengxiong-ruan commented on pkg/sql/schemachanger/scbuild/internal/scbuildstmt/dependencies.go:
```diff
@@ -325,3 +358,15 @@ type NameResolver interface {
+type ReferenceProvider interface {
+	// ForEachTableReference iterate through all referenced tables and the
+	// reference details with the given function.
+	ForEachTableReference(func(tblID descpb.ID, refs []descpb.TableDescriptor_Reference) error) error
```
Done. Thanks for the suggestion. I made the interface into more detailed methods so that we don't need to reason about things like descriptor types (view, sequence, type, table) anymore.

chengxiong-ruan commented on pkg/sql/schemachanger/scdeps/sctestdeps/test_state.go:
```diff
@@ -227,6 +229,14 @@ func (s *TestState) ClientNoticeSender() eval.ClientNoticeSender {
 	return &faketreeeval.DummyClientNoticeSender{}
 }
 
+func (s *TestState) DescIDGenerator() eval.DescIDGenerator {
```
Done

chengxiong-ruan commented on pkg/sql/schemachanger/scexec/exec_immediate_mutation.go:
```diff
@@ -25,6 +25,7 @@ type immediateState struct {
 	drainedNames        map[descpb.ID][]descpb.NameInfo
 	descriptorsToDelete catalog.DescriptorIDSet
 	commentsToUpdate    []commentToUpdate
+	newDescriptors      nstree.IDMap
```
Good point, made it a map.

chengxiong-ruan commented on pkg/sql/schemachanger/scplan/internal/opgen/opgen_function_body.go:
```diff
@@ -21,8 +21,24 @@ func init() {
-					return notImplemented(this)
+				emit(func(this *scpb.FunctionBody) *scop.SetFunctionBody {
+					return &scop.SetFunctionBody{
+						Body: *this,
```
Done.

chengxiong-ruan commented on pkg/sql/schemachanger/scplan/internal/opgen/opgen_object_parent.go:
```diff
@@ -20,8 +20,8 @@ func init() {
-				emit(func(this *scpb.ObjectParent) *scop.NotImplemented {
-					return notImplemented(this)
+				emit(func(this *scpb.ObjectParent) *scop.SetObjectParentID {
+					return &scop.SetObjectParentID{ObjParent: *this}
```
done.

chengxiong-ruan commented on pkg/sql/schemachanger/scplan/internal/rules/current/dep_create_function.go:
```diff
@@ -0,0 +1,36 @@
+// Copyright 2022 The Cockroach Authors.
```
Fixed all.

chengxiong-ruan commented on pkg/sql/schemachanger/scplan/internal/rules/current/helpers.go:
```diff
@@ -85,6 +85,10 @@ func isSimpleDependent(e scpb.Element) bool {
 
+func isDependent(e scpb.Element) bool {
+	return !isDescriptor(e)
+}
```
Done.

chengxiong-ruan commented on pkg/sql/schemachanger/scbuild/build.go:
```diff
@@ -110,7 +110,9 @@ func Build(
-		if desc.HasConcurrentSchemaChanges() {
+		// If a descriptor is being created, we don't need to worry about concurrent
+		// schema changes.
+		if desc != nil && desc.HasConcurrentSchemaChanges() {
```
Good point, Done.

chengxiong-ruan reviewed (commented):
Thanks for the review! I've done all you suggested.

chengxiong-ruan commented:
TFTRs!
bors r+

postamar commented:
Thanks for doing this!
